drivers/sensor/ds18b20/ds18b20.c

```diff
@@ -0,0 +1,256 @@
+
+	val->val1 = (temp & 0x0FFF) >> 4;
+	val->val2 = (temp & 0x000F) * 10000;
+}
+
+/*
+ * Write scratch pad, read back, then copy to eeprom
+ */
+static int ds18b20_write_scratchpad(const struct device *dev,
+				    struct ds18b20_scratchpad scratchpad)
+{
+
+	struct ds18b20_data *data = dev->data;
+	const struct device *bus = ds18b20_bus(dev);
+	uint8_t sp_data[3] = {
+		scratchpad.alarm_temp_high,
+		scratchpad.alarm_temp_low,
+		scratchpad.config
+	};
+
+	w1_reset_select(bus, &data->rom);
+	w1_write_byte(bus, DS18B20_CMD_WRITE_SCRATCHPAD);
+	w1_write_block(bus, &sp_data[0], sizeof(sp_data));
+
+	return 0;
+}
+
+static int ds18b20_read_scratchpad(const struct device *dev,
+				   struct ds18b20_scratchpad *scratchpad)
+{
+	struct ds18b20_data *data = dev->data;
+	const struct device *bus = ds18b20_bus(dev);
+
+	w1_reset_select(bus, &data->rom);
+	w1_write_byte(bus, DS18B20_CMD_READ_SCRATCHPAD);
+	w1_read_block(bus, (uint8_t *)&scratchpad[0], 9);
+
+	return 0;
+}
+
+/* Starts sensor temperature conversion without waiting for completion. */
+static void ds18b20_temperature_convert(const struct device *dev)
+{
+	struct ds18b20_data *data = dev->data;
+	const struct device *bus = ds18b20_bus(dev);
+
+	w1_reset_select(bus, &data->rom);
+	w1_write_byte(bus, DS18B20_CMD_CONVERT_T);
+}
+
+/* Read power supply status from the sensor. */
+static int ds18b20_read_power_supply(const struct device *dev)
+{
+	struct ds18b20_data *data = dev->data;
+	const struct device *bus = ds18b20_bus(dev);
+
+	w1_reset_select(bus, &data->rom);
+	w1_write_byte(bus, DS18B20_CMD_READ_POWER_SUPPLY);
+	data->power_supply = w1_read_bit(bus);
+
+	return 0;
+}
+
+/*
+ * Write resolution into configuration struct,
+ * but don't write it to the sensor yet.
+ */
+static int ds18b20_set_resolution(const struct device *dev, uint8_t resolution)
+{
+	struct ds18b20_data *data = dev->data;
+
+	data->scratchpad.config &= ~DS18B20_RESOLUTION_MASK;
+	data->scratchpad.config |= DS18B20_RESOLUTION(resolution);
+
+	return 0;
+}
+
+static int ds18b20_sample_fetch(const struct device *dev,
+				enum sensor_channel chan)
+{
+	const struct ds18b20_config *cfg = dev->config;
+	struct ds18b20_data *data = dev->data;
+	int status = 0;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL ||
+			chan == SENSOR_CHAN_AMBIENT_TEMP);
+
+	if (!data->lazy_loaded) {
+		status = ds18b20_configure(dev);
+
+		if (status < 0) {
+			return status;
+		}
+		data->lazy_loaded = true;
+	}
+
+	ds18b20_temperature_convert(dev);
+	k_sleep(K_MSEC(measure_wait_ms[
+				DS18B20_RESOLUTION_INDEX(cfg->resolution)]));
+	ds18b20_read_scratchpad(dev, &data->scratchpad);
+
+	return 0;
+}
+
+static int ds18b20_channel_get(const struct device *dev,
+			       enum sensor_channel chan,
+			       struct sensor_value *val)
+{
+	struct ds18b20_data *data = dev->data;
+
+	if (chan == SENSOR_CHAN_AMBIENT_TEMP) {
+		ds18b20_temperature_from_raw((uint8_t *)&data->scratchpad.temp,
+					     val);
+	} else {
+		return -ENOTSUP;
+	}
+
+	return 0;
```

Comment:
Should return `rc`, right? Otherwise it would return 0 even if `rc` is set to `-ENOTSUP`.
```suggestion
	return rc;
```

Reply:
Fixed.